File: jysmol_lua/src/stringifier.lua
local M = {}

local function is_array(t)
  local i = 0
  for key in pairs(t) do
      i = i + 1
      if type(key) ~= 'number' then
          return false
      end
  end
  return true
end

function M.stringify_string(value)
    return '"' .. value .. '"'
end

function M.stringify_number(value)
    return '' .. value .. ''
end

function M.stringify_array(arr, stack)
    local str = '['

    if stack[arr] then
        error('circular reference error')
    end

    stack[arr] = true

    for _, el in ipairs(arr) do
        str = str .. M.stringify_value(el) .. ','
    end

    str = str .. ']'

    stack[arr] = nil

    return str
end

function M.stringify_object(obj, stack)
    local str = '{'

    if stack[obj] then
        error('circular reference error')
    end

    stack[obj] = true

    for key, value in pairs(obj) do
        str = str .. M.stringify_string(key) .. ':' .. M.stringify_value(value) .. ','
    end

    str = str .. '}'

    stack[obj] = nil

    return str
end

function M.stringify_value(value, stack)
    if value == nil then return 'null' -- TODO useless in table
    elseif value == true then return 'true'
    elseif value == false then return 'false'
    end

    local val_type = type(value)

    if val_type == 'string' then return M.stringify_string(value)
    elseif val_type == 'number' then return M.stringify_number(value)
    elseif val_type == 'table' then
        if is_array(value) then
            return M.stringify_array(value, stack or {})
        end
        return M.stringify_object(value, stack or {})
    end

    error('unsupported type' .. val_type)
end

return M

File: jysmol_python/stringifier.py
def stringify_value(value, stack):
    match value:
        case None: return 'null'
        case False: return 'false'
        case True: return 'true'

    value_t = type(value) 

    if (value_t is int or value_t is float): return stringify_number(value)
    elif (value_t is str): return stringify_string(value)
    elif (value_t is list): return stringify_array(value, stack or set())
    elif (value_t is dict): return stringify_object(value, stack or set())

    raise Exception('nie ma')

def stringify_string(value):
    return f'"{value}"'

def stringify_number(value):
    return f'{value}'

def stringify_array(arr, stack):
    out = '['
    arr_id = id(arr)

    if (arr_id in stack):
        raise Exception('circular reference')

    stack.add(arr_id)

    for el in arr:
        out += stringify_value(el, stack)
        out += ','

    out += ']'

    stack.remove(arr_id)

    return out

def stringify_object(obj, stack):
    out = '{'
    obj_id = id(obj)

    if (obj_id in stack):
        raise Exception('circular reference')

    stack.add(obj_id)

    for key, value in obj.items():
        out += stringify_string(key)
        out += ':'
        out += stringify_value(value, stack)
        out += ','

    out += '}'

    stack.remove(obj_id)

    return out
